File: apps/api/src/s3.rs
use aws_sdk_s3::{
    config::{Credentials, Region},
    Client, Config,
};

use crate::config::Config as AppConfig;

#[derive(Clone)]
pub struct S3Client {
    pub client: Client,
    pub bucket: String,
}

impl S3Client {
    pub async fn new(config: &AppConfig) -> Self {
        let credentials = Credentials::new(
            &config.aws_access_key_id,
            &config.aws_secret_access_key,
            None,
            None,
            "s3",
        );

        let s3_config = Config::builder()
            .endpoint_url(&config.aws_endpoint_url)
            .region(Region::new("auto"))
            .credentials_provider(credentials)
            .force_path_style(true)
            .build();

        let client = Client::from_conf(s3_config);

        tracing::info!("S3 client initialized for bucket: {}", config.aws_s3_bucket_name);

        Self {
            client,
            bucket: config.aws_s3_bucket_name.clone(),
        }
    }

    pub fn get_repo_prefix(user_id: &str, repo_name: &str) -> String {
        format!("repos/{}/{}.git", user_id, repo_name)
    }

    pub async fn get_object(&self, key: &str) -> Option<Vec<u8>> {
        let result = self.client
            .get_object()
            .bucket(&self.bucket)
            .key(key)
            .send()
            .await;

        match result {
            Ok(output) => {
                output.body.collect().await.ok().map(|data| data.to_vec())
            }
            Err(_) => None,
        }
    }

    pub async fn put_object(&self, key: &str, data: Vec<u8>) -> Result<(), aws_sdk_s3::Error> {
        self.client
            .put_object()
            .bucket(&self.bucket)
            .key(key)
            .body(data.into())
            .send()
            .await?;
        Ok(())
    }

    pub async fn delete_object(&self, key: &str) -> Result<(), aws_sdk_s3::Error> {
        self.client
            .delete_object()
            .bucket(&self.bucket)
            .key(key)
            .send()
            .await?;
        Ok(())
    }

    pub async fn list_objects(&self, prefix: &str) -> Vec<String> {
        let mut keys = Vec::new();
        let mut continuation_token: Option<String> = None;

        loop {
            let mut request = self.client
                .list_objects_v2()
                .bucket(&self.bucket)
                .prefix(prefix);

            if let Some(token) = continuation_token.take() {
                request = request.continuation_token(token);
            }

            match request.send().await {
                Ok(output) => {
                    if let Some(contents) = output.contents {
                        for obj in contents {
                            if let Some(key) = obj.key {
                                keys.push(key);
                            }
                        }
                    }

                    if output.is_truncated.unwrap_or(false) {
                        continuation_token = output.next_continuation_token;
                    } else {
                        break;
                    }
                }
                Err(_) => break,
            }
        }

        keys
    }

    pub async fn delete_prefix(&self, prefix: &str) -> Result<(), aws_sdk_s3::Error> {
        let keys = self.list_objects(prefix).await;

        for chunk in keys.chunks(1000) {
            let objects: Vec<_> = chunk
                .iter()
                .map(|key| {
                    aws_sdk_s3::types::ObjectIdentifier::builder()
                        .key(key)
                        .build()
                        .unwrap()
                })
                .collect();

            if !objects.is_empty() {
                self.client
                    .delete_objects()
                    .bucket(&self.bucket)
                    .delete(
                        aws_sdk_s3::types::Delete::builder()
                            .set_objects(Some(objects))
                            .build()
                            .unwrap()
                    )
                    .send()
                    .await?;
            }
        }

        Ok(())
    }
}

File: apps/api/src/s3.ts
import { S3Client, GetObjectCommand, PutObjectCommand, DeleteObjectCommand, ListObjectsV2Command, HeadObjectCommand } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";
import { config } from "./config";

const s3Configured = Boolean(
  config.s3.endpoint && config.s3.region && config.s3.bucket && config.s3.accessKeyId && config.s3.secretAccessKey
);

export const s3Client = s3Configured
  ? new S3Client({
    endpoint: config.s3.endpoint,
    region: config.s3.region,
    credentials: {
      accessKeyId: config.s3.accessKeyId,
      secretAccessKey: config.s3.secretAccessKey,
    },
    forcePathStyle: true,
  })
  : null;

export const bucket = config.s3.bucket;

export const getRepoPrefix = (owner: string, repo: string): string => {
  return `repos/${owner}/${repo}`;
};

export const getObject = async (key: string): Promise<Buffer | null> => {
  if (!s3Client) {
    return null;
  }
  try {
    const response = await s3Client.send(
      new GetObjectCommand({
        Bucket: bucket,
        Key: key,
      })
    );

    if (!response.Body) {
      return null;
    }

    const bytes = await response.Body.transformToByteArray();
    return Buffer.from(bytes);
  } catch (error: any) {
    if (error.name === "NoSuchKey" || error.$metadata?.httpStatusCode === 404) {
      return null;
    }
    throw error;
  }
};

export const putObject = async (key: string, body: Buffer | Uint8Array | string, contentType?: string): Promise<void> => {
  if (!s3Client) {
    throw new Error("S3 is not configured");
  }
  await s3Client.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: body,
      ContentType: contentType,
    })
  );
};

export const deleteObject = async (key: string): Promise<void> => {
  if (!s3Client) {
    throw new Error("S3 is not configured");
  }
  await s3Client.send(
    new DeleteObjectCommand({
      Bucket: bucket,
      Key: key,
    })
  );
};

export const listObjects = async (prefix: string): Promise<string[]> => {
  if (!s3Client) {
    return [];
  }
  const keys: string[] = [];
  let continuationToken: string | undefined;

  do {
    const response = await s3Client.send(
      new ListObjectsV2Command({
        Bucket: bucket,
        Prefix: prefix,
        ContinuationToken: continuationToken,
      })
    );

    if (response.Contents) {
      for (const obj of response.Contents) {
        if (obj.Key) {
          keys.push(obj.Key);
        }
      }
    }

    continuationToken = response.NextContinuationToken;
  } while (continuationToken);

  return keys;
};

export const objectExists = async (key: string): Promise<boolean> => {
  if (!s3Client) {
    return false;
  }
  try {
    await s3Client.send(
      new HeadObjectCommand({
        Bucket: bucket,
        Key: key,
      })
    );
    return true;
  } catch (error: any) {
    if (error.name === "NotFound" || error.$metadata?.httpStatusCode === 404) {
      return false;
    }
    throw error;
  }
};

export const deletePrefix = async (prefix: string): Promise<void> => {
  if (!s3Client) {
    throw new Error("S3 is not configured");
  }
  const keys = await listObjects(prefix);
  for (const key of keys) {
    await deleteObject(key);
  }
};

export const uploadMultipart = async (
  key: string,
  body: Buffer | Uint8Array | ReadableStream,
  contentType?: string
): Promise<void> => {
  if (!s3Client) {
    throw new Error("S3 is not configured");
  }
  const upload = new Upload({
    client: s3Client,
    params: {
      Bucket: bucket,
      Key: key,
      Body: body,
      ContentType: contentType,
    },
  });

  await upload.done();
};

export const getObjectStream = async (key: string): Promise<ReadableStream | null> => {
  if (!s3Client) {
    return null;
  }
  try {
    const response = await s3Client.send(
      new GetObjectCommand({
        Bucket: bucket,
        Key: key,
      })
    );

    if (!response.Body) {
      return null;
    }

    return response.Body.transformToWebStream();
  } catch (error: any) {
    if (error.name === "NoSuchKey" || error.$metadata?.httpStatusCode === 404) {
      return null;
    }
    throw error;
  }
};

export const copyPrefix = async (sourcePrefix: string, targetPrefix: string): Promise<void> => {
  const keys = await listObjects(sourcePrefix);
  const normalizedSource = sourcePrefix.replace(/\/$/, "");
  const normalizedTarget = targetPrefix.replace(/\/$/, "");

  for (const key of keys) {
    const data = await getObject(key);
    if (!data) {
      continue;
    }
    const suffix = key.slice(normalizedSource.length);
    const targetKey = `${normalizedTarget}${suffix}`;
    await putObject(targetKey, data);
  }
};
